Update `yahnis-elsts/plugin-update-checker` from 4.13 to 5.1

dkotter opened the pull request:
### Description of the Change

The Plugin Update Checker library we use is also used on a few other 10up projects.  We need to ensure we are using the same version across all of these or you can end up with conflicts (especially when using composer to manage all dependencies of a project). This PR updates this library to the latest version, which will also be done in other repos that are using this library.

### How to test the Change

1. Pull these changes down
2. Within the `classifai.php` file, set the version number to something lower, like 2.1.0
3. Ensure you have a proper license in place
4. Go to the plugins section and ensure you see an update notice for Classifai (may need to click the `Check for updates` link; also ensure the plugin is active)
5. Click on the `View version X.X.X details` link and ensure you see proper details
6. Go to the WordPress Updates page and ensure you see Classifai there as well, with the icon
7. Click on the `View version X.X.X details` link and ensure you see proper details
8. If desired, click the `update now` link and ensure an update works (either from the plugins page or the main updates page). Note though that this will wipe out your version controlled `classifai` directory, so proceed with caution

### Changelog Entry

> Changed - Upgrade the Plugin Update Checker library to 5.1

### Credits

Props @dkotter

### Checklist:
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

dkotter committed:
Update our plugin checker to v5.1

Sidsector9 commented:
I'm having a bit trouble reproducing step (4). I'll investigate it tomorrow.

dkotter commented:
> I'm having a bit trouble reproducing step (4). I'll investigate it tomorrow.

If you're on multisite, you'll need to go to the network plugins screen to see updates (this caught me when testing). If you have caching in place, that may also play a role and will need to be flushed

ravinderk commented:
> I'm having a bit of trouble reproducing step (4). I'll investigate it tomorrow.

@Sidsector9 I also face the same issue. I resolved it by re-installing composer packages and node packages.

ravinderk reviewed (approved):
@dkotter It is working as expected.


<img width="665" alt="image" src="https://github.com/10up/classifai/assets/1784821/33c99c6d-1413-49d1-8318-9b6f030e8119">
<img width="977" alt="image" src="https://github.com/10up/classifai/assets/1784821/6a8c88cf-79fe-47b1-85cd-2245b7434db8">
<img width="948" alt="image" src="https://github.com/10up/classifai/assets/1784821/92ac761d-514c-4e04-a611-512d590e5ddd">

ravinderk commented on includes/Classifai/Admin/Update.php:
```diff
@@ -7,7 +7,7 @@
 namespace Classifai\Admin;
 
-use Puc_v4_Factory;
+use YahnisElsts\PluginUpdateChecker\v5\PucFactory;
```
@dkotter This can cause package conflict between two plugins if both use different versions. I would suggest adding a custom prefix to vendor packages which we need in the product environment.

dkotter replied:
So we originally decided not to use a tool like PHP Scoper / Mozart / strauss (see https://github.com/10up/classifai/pull/365) and just ensure we were using the same version across our plugins. I'm not sure adding namespacing would solve the issue we've run into so far, where if someone is using composer to manage their plugins, it won't allow you to install multiple versions of the same dependency. That said, I'm not opposed to adding namespacing here though I won't have time to look into this until next week and ideally we were hoping to get this change out in the next release (tentatively scheduled for sometime next week). I know you've done work with `strauss` in the past, if you have interest and time to look at integrating that (or something similar here) feel free to takeover this PR

ravinderk replied:
@dkotter I will create an issue and resolve it separately pr.

ravinderk reviewed (changes requested):
@dkotter I noticed the following error in debug.log. Can you check?

```log
[08-Aug-2023 06:50:09 UTC] PHP Notice:  Trying to get property 'filename' of non-object in /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-content/plugins/classifai/includes/Classifai/Admin/Update.php on line 91
[08-Aug-2023 06:50:09 UTC] PHP Stack trace:
[08-Aug-2023 06:50:09 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-admin/update-core.php:0
[08-Aug-2023 06:50:09 UTC] PHP   2. require_once() /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-admin/update-core.php:10
[08-Aug-2023 06:50:09 UTC] PHP   3. do_action($hook_name = 'load-update-core.php') /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-admin/admin.php:385
[08-Aug-2023 06:50:09 UTC] PHP   4. WP_Hook->do_action($args = [0 => '']) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/plugin.php:517
[08-Aug-2023 06:50:09 UTC] PHP   5. WP_Hook->apply_filters($value = '', $args = [0 => '']) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/class-wp-hook.php:334
[08-Aug-2023 06:50:09 UTC] PHP   6. wp_update_plugins($extra_stats = '') /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/class-wp-hook.php:310
[08-Aug-2023 06:50:09 UTC] PHP   7. set_site_transient($transient = 'update_plugins', $value = class stdClass { public $last_checked = 1691477409; public $response = ['autoshare-for-twitter/autoshare-for-twitter.php' => class stdClass { ... }, 'custom-post-type-ui/custom-post-type-ui.php' => class stdClass { ... }, 'polylang/polylang.php' => class stdClass { ... }]; public $translations = []; public $no_update = ['akismet/akismet.php' => class stdClass { ... }, 'classic-editor/classic-editor.php' => class stdClass { ... }, 'insert-special-characters/insert-special-characters.php' => class stdClass { ... }, 'loco-translate/loco.php' => class stdClass { ... }, 'safe-redirect-manager/safe-redirect-manager.php' => class stdClass { ... }, 'safe-svg/safe-svg.php' => class stdClass { ... }, 'simple-page-ordering/simple-page-ordering.php' => class stdClass { ... }, 'classifai/classifai.php' => class stdClass { ... }]; public $checked = ['10up-sitemaps/10up-sitemaps.php' => '1.2', 'akismet/akismet.php' => '5.2', 'autoshare-for-twitter/autoshare-for-twitter.php' => '2.0.0', 'classic-editor/classic-editor.php' => '1.6.3', 'classifai/classifai.php' => '2.3.0-dev', 'custom-post-type-ui/custom-post-type-ui.php' => '1.13.5', 'distributor/distributor.php' => '2.0.0-rc1', 'auto-distribute/auto-distribute.php' => '0.2.0', 'distributor-edit-remote-post-meta/plugin.php' => '0.2.1', 'insert-special-characters/insert-special-characters.php' => '1.0.7', 'loco-translate/loco.php' => '2.6.4', 'mention-notifier/plugin.php' => '0.1.0', 'polylang/polylang.php' => '3.3.3', 'safe-redirect-manager/safe-redirect-manager.php' => '2.0.1', 'safe-svg/safe-svg.php' => '2.1.1', 'simple-page-ordering/simple-page-ordering.php' => '2.5.1', 'speedyseo/speedyseo.php' => '0.2.0'] }, $expiration = *uninitialized*) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/update.php:393
[08-Aug-2023 06:50:09 UTC] PHP   8. apply_filters($hook_name = 'pre_set_site_transient_update_plugins', $value = class stdClass { public $last_checked = 1691477409; public $response = ['autoshare-for-twitter/autoshare-for-twitter.php' => class stdClass { ... }, 'custom-post-type-ui/custom-post-type-ui.php' => class stdClass { ... }, 'polylang/polylang.php' => class stdClass { ... }]; public $translations = []; public $no_update = ['akismet/akismet.php' => class stdClass { ... }, 'classic-editor/classic-editor.php' => class stdClass { ... }, 'insert-special-characters/insert-special-characters.php' => class stdClass { ... }, 'loco-translate/loco.php' => class stdClass { ... }, 'safe-redirect-manager/safe-redirect-manager.php' => class stdClass { ... }, 'safe-svg/safe-svg.php' => class stdClass { ... }, 'simple-page-ordering/simple-page-ordering.php' => class stdClass { ... }, 'classifai/classifai.php' => class stdClass { ... }]; public $checked = ['10up-sitemaps/10up-sitemaps.php' => '1.2', 'akismet/akismet.php' => '5.2', 'autoshare-for-twitter/autoshare-for-twitter.php' => '2.0.0', 'classic-editor/classic-editor.php' => '1.6.3', 'classifai/classifai.php' => '2.3.0-dev', 'custom-post-type-ui/custom-post-type-ui.php' => '1.13.5', 'distributor/distributor.php' => '2.0.0-rc1', 'auto-distribute/auto-distribute.php' => '0.2.0', 'distributor-edit-remote-post-meta/plugin.php' => '0.2.1', 'insert-special-characters/insert-special-characters.php' => '1.0.7', 'loco-translate/loco.php' => '2.6.4', 'mention-notifier/plugin.php' => '0.1.0', 'polylang/polylang.php' => '3.3.3', 'safe-redirect-manager/safe-redirect-manager.php' => '2.0.1', 'safe-svg/safe-svg.php' => '2.1.1', 'simple-page-ordering/simple-page-ordering.php' => '2.5.1', 'speedyseo/speedyseo.php' => '0.2.0'] }, ...$args = variadic('update_plugins')) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/option.php:2053
[08-Aug-2023 06:50:09 UTC] PHP   9. WP_Hook->apply_filters($value = class stdClass { public $last_checked = 1691477409; public $response = ['autoshare-for-twitter/autoshare-for-twitter.php' => class stdClass { ... }, 'custom-post-type-ui/custom-post-type-ui.php' => class stdClass { ... }, 'polylang/polylang.php' => class stdClass { ... }]; public $translations = []; public $no_update = ['akismet/akismet.php' => class stdClass { ... }, 'classic-editor/classic-editor.php' => class stdClass { ... }, 'insert-special-characters/insert-special-characters.php' => class stdClass { ... }, 'loco-translate/loco.php' => class stdClass { ... }, 'safe-redirect-manager/safe-redirect-manager.php' => class stdClass { ... }, 'safe-svg/safe-svg.php' => class stdClass { ... }, 'simple-page-ordering/simple-page-ordering.php' => class stdClass { ... }, 'classifai/classifai.php' => class stdClass { ... }]; public $checked = ['10up-sitemaps/10up-sitemaps.php' => '1.2', 'akismet/akismet.php' => '5.2', 'autoshare-for-twitter/autoshare-for-twitter.php' => '2.0.0', 'classic-editor/classic-editor.php' => '1.6.3', 'classifai/classifai.php' => '2.3.0-dev', 'custom-post-type-ui/custom-post-type-ui.php' => '1.13.5', 'distributor/distributor.php' => '2.0.0-rc1', 'auto-distribute/auto-distribute.php' => '0.2.0', 'distributor-edit-remote-post-meta/plugin.php' => '0.2.1', 'insert-special-characters/insert-special-characters.php' => '1.0.7', 'loco-translate/loco.php' => '2.6.4', 'mention-notifier/plugin.php' => '0.1.0', 'polylang/polylang.php' => '3.3.3', 'safe-redirect-manager/safe-redirect-manager.php' => '2.0.1', 'safe-svg/safe-svg.php' => '2.1.1', 'simple-page-ordering/simple-page-ordering.php' => '2.5.1', 'speedyseo/speedyseo.php' => '0.2.0'] }, $args = [0 => class stdClass { public $last_checked = 1691477409; public $response = [...]; public $translations = [...]; public $no_update = [...]; public $checked = [...] }, 1 => 'update_plugins']) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/plugin.php:205
[08-Aug-2023 06:50:09 UTC] PHP  10. Classifai\Admin\Update->maybe_update($transient = class stdClass { public $last_checked = 1691477409; public $response = ['autoshare-for-twitter/autoshare-for-twitter.php' => class stdClass { ... }, 'custom-post-type-ui/custom-post-type-ui.php' => class stdClass { ... }, 'polylang/polylang.php' => class stdClass { ... }]; public $translations = []; public $no_update = ['akismet/akismet.php' => class stdClass { ... }, 'classic-editor/classic-editor.php' => class stdClass { ... }, 'insert-special-characters/insert-special-characters.php' => class stdClass { ... }, 'loco-translate/loco.php' => class stdClass { ... }, 'safe-redirect-manager/safe-redirect-manager.php' => class stdClass { ... }, 'safe-svg/safe-svg.php' => class stdClass { ... }, 'simple-page-ordering/simple-page-ordering.php' => class stdClass { ... }, 'classifai/classifai.php' => class stdClass { ... }]; public $checked = ['10up-sitemaps/10up-sitemaps.php' => '1.2', 'akismet/akismet.php' => '5.2', 'autoshare-for-twitter/autoshare-for-twitter.php' => '2.0.0', 'classic-editor/classic-editor.php' => '1.6.3', 'classifai/classifai.php' => '2.3.0-dev', 'custom-post-type-ui/custom-post-type-ui.php' => '1.13.5', 'distributor/distributor.php' => '2.0.0-rc1', 'auto-distribute/auto-distribute.php' => '0.2.0', 'distributor-edit-remote-post-meta/plugin.php' => '0.2.1', 'insert-special-characters/insert-special-characters.php' => '1.0.7', 'loco-translate/loco.php' => '2.6.4', 'mention-notifier/plugin.php' => '0.1.0', 'polylang/polylang.php' => '3.3.3', 'safe-redirect-manager/safe-redirect-manager.php' => '2.0.1', 'safe-svg/safe-svg.php' => '2.1.1', 'simple-page-ordering/simple-page-ordering.php' => '2.5.1', 'speedyseo/speedyseo.php' => '0.2.0'] }) /Users/ravinderkumar/Local-Sites/10osp/app/public/wp-includes/class-wp-hook.php:312

```

-----------------------------------------
**Step to reproduce**

This issue only happens once when you switch from `develop` to `feature/update-plugin-checker`.

- Switch to `develop`
- Go to the WordPress Updates page.
- Switch to `feature/update-plugin-checker`
- Go to the WordPress Updates page.
- Check `debug.log`

dkotter committed:
Merge branch 'develop' into feature/update-plugin-checker

dkotter committed:
Ensure we have a proper update object before using it

dkotter commented:
> @dkotter I noticed the following error in debug.log. Can you check?

I was able to reproduce this but couldn't figure out why it's happening. I added a check to resolve this issue though would be great to understand what changed to cause this. As far as I can tell, everything still works as expected so this may not be a big deal.

In addition, the method that was causing this error is the `maybe_update` method. I'm not sure this is even needed, as I removed this and things still seemed to work. In looking at the [documentation](https://github.com/YahnisElsts/plugin-update-checker) for the update checker, I'm not even seeing this code referenced there so not quite sure why we're using it (seems it was copied from Distributor). But may be safer to just leave for now
